Add noInterop option to babel-plugin-transform-es2015-modules-commonjs.

rwjblue opened the pull request:
The intent of this option is to toggle module interop behavior. When `true`
the `interopRequireXXX` helper invocations will not be emitted in the transpiled output.


<details>
  <summary>Original PR Description</summary>
The intent of the `strict` option for `babel-plugin-transform-es2015-modules-commonjs` was to toggle module interop behavior, 
however a few scenarios were not being handled properly.

Specifically, when only a default import is used from a given module, such as:

```js
import Foo from 'foo';
```

The transpiled output with and without `strict` flag still included `_interopRequireDefault`.

After this change, the output is now properly disabling the inclusion of `_interopRequireDefault`.
</details>
Fixes https://github.com/babel/babel/issues/4073.

rwjblue commented:
I am pushing very hard on getting ember-cli to utilize babel 6 (yes, I know it has been way to long 😢), and this is a fairly large blocker for us to migrate to 6.  The reason for this, is that we have been using `amdStrict` throughout the ecosystem (because we build only for the browser and do not have a commonjs interop requirement).  Now that we are migrating to babel@6, we need to ensure that modules transpiled by babel@5 with `amdStrict` **and** modules transpiled by babel@6 can interop with each other. 

Are there any opposition to backporting this to babel@6 to unblock the ember-cli migration efforts?

/cc @stefanpenner @hzoo

rwjblue commented:
FWIW, I'm happy to decouple the options as suggested in https://github.com/babel/babel/issues/4071 for babel@7 as a follow-up to this PR if y'all are 👍 on that...

loganfsmyth commented:
> The intent of the strict option for babel-plugin-transform-es2015-modules-commonjs was to toggle module interop behavior

I honestly can't say what the original intent behind it was in 6.x. In Babel 6 that has not been the case, whether on purpose or on accident, but I'd be extremely hesitant to change this behavior.

Perhaps we can add a "noInterop" option and deprecate usage of `strict`? `strict` isn't a very clear name anyway.

rwjblue commented:
My statement above came from your comment/issue detail in https://github.com/babel/babel/issues/4071. I'm happy to just implement the suggestion from https://github.com/babel/babel/issues/4071 (adding `strictMode` and either `strictModules`, `noInterop` or something). I was unsure if that would be acceptable for backporting though. 

If that is preferable, I can do that here. My primary goal is just to find a path forward for ember-cli ecosystem.

loganfsmyth commented:
Ahh, I guess I did write that :P I think that was in reference to the presence of `__esModule` only, I don't think I intended to imply that it was about `interopRequire`, but since it's been a year and I wasn't very specific it is hard to say.

We did end up separating `strict` and `strictMode` so that should be set. As for `strictModules`, looking at it now I'd in favor of a more explicit name like `noInterop` I think.

One question for you, is there anything gained by having `noInterop` leave out `__esModule` like `strict` does now, for your usecase? I'd almost be tempted to have it leave that in and just drop the usage of `interopRequire` based on the flag separately.

rwjblue commented:
Yes, generally speaking I think I would prefer to keep `__esModule` and have `noInterop` only control the usage of the various interop require statements. 

I'll get things updated here tonight, and let y'all know when it's ready for another review.

rwjblue commented:
I just pushed the update as discussed, please review. I am working on updating README's and threading the new option through the various presets.

---

While working on it, I noticed that the `strict` option is pretty bizarre at the moment.  It currently does:

* Prevent usage of `interopRequireWildcard` [here](https://github.com/babel/babel/blob/8a82cc060ae0ab46bf52e05e592de770bd246f6f/packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js#L374)
* Prevent definition of `exports.__esModule` [here](https://github.com/babel/babel/blob/8a82cc060ae0ab46bf52e05e592de770bd246f6f/packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js#L466)

From the conversation above, I _think_ the first (disabling `interopRequireWildcard`) should not actually be part of what `strict` is responsible for.  However, changing it may actually break folks somehow relying on it.  At this point I have left both of these behaviors intact, but if you'd like I can "fix" `strict` to only toggle the `__esModule` definition.

rwjblue commented:
OK, updated README for `es2015-modules-commonjs` to include `strict` and `noInterop` in the docs there.

I decided not to update the various presets to allow this to be passed through, since that seems to be outside the norm for the presets (there is currently no way to pass through `strict` or `allowTopLevelThis` which are valid options for `es2015-modules-commonjs`).  If someone needs to customize more than just `loose`, they should probably use `modules: false` and add the modules plugin themselves with their desired configuration.  Happy to tweak somehow If this assumption is incorrect.

---

I believe that I have addressed the requests from the prior review, and this is ready for another peek...

loganfsmyth commented:
Nice. Looks good to me. If you'd like to backport to `master` feel free to post another PR.
